### PR TITLE
feat: [rfc]Add user trace cli for mme logs

### PR DESCRIPTION
**Pull request by wallyrb**
Signed-off-by: Wally Rodriguez B <wallyrb@fb.com>

## Summary

User trace cli should facilitate filtering user control sessions in mme logs. The cli has 3 functionalities for a given IMSI and mme log.

- 'list_imsi'
Provide a list of imsi and number of occurrences found in logs. (No input required)

- 'list_ue_id'
List mme and enb ue id pairs for a given IMSI. (IMSI input required)

- 'session_trace'
Output session trace for given user id(mme_id) and IMSI (IMSI and mme_id required). If is a known issue, it will provide the error and suggestions. Two known errors have been added(APN error and IE not supported).
I use regex to match different ue_id patterns.

Additional inputs

-p : "Path to the data directory"
-i: "imsi"
-m: "mme_id"

In general, we can use below flow for this cli.

1. Grab one imsi from list_imsi
2. From the IMSI selected get the ue_id found from list_ue_id
3. Output with session_trace using imsi + mme_id found in Step 1 and 2.

Examples: 
```
python3 user_trace_cli.py -p mme_apn.log list_ue_id -i 000000000000110 

IMSI: 000000000000110

mme_id          enodeb_id        
-------------------------
25 0x19         728 0x2d8            
26 0x1a         729 0x2d9            
27 0x1b         730 0x2da            
28 0x1c         731 0x2db            
29 0x1d         732 0x2dc   
```

```
python3 user_trace_cli.py -p mme_ie.log list_imsi                      

IMSI            Occurrences     
------------------------------
000000000000110 331            
000000000000111 2              
000000000000112 2              
000000000000113 2              
000000000000114 2              
000000000000115 2              
000000000000116 2              
000000000000117 2              
000000000000118 2              
000000000000119 2      
```

For session trace logs, please find in https://www.internalfb.com/phabricator/paste/view/P461580010

## Test Plan

list_ue_id: https://www.internalfb.com/phabricator/paste/view/P461580714
list_imsi: https://www.internalfb.com/phabricator/paste/view/P461580340
session_trace: https://www.internalfb.com/phabricator/paste/view/P461580010

## Additional Information

Documentation added

image

### CODE REVIEW

**Comment by pshelar** [comment]
can we read state from redis rather than from mme logs?

**Comment by wallyrb** [comment]
> can we read state from redis rather than from mme logs?

@pshelar  thanks for checking. I believe mme logs provide more detail information for troubleshooting control plane user issues. I think Redis will provide a snapshot of the last state of the user. The purpose of cli is to provide the journey of control messaging for a specific user, and if there was a failure, what messages were found in the middle.  

For example session_trace:  https://www.internalfb.com/phabricator/paste/view/P461580010

The option "list_ue_id" will provide historically which ue id were found for this user in the logs.(how many "sessions"). Not sure if redis provides historical view, I believe only the last state(last mme, enodeb id)

- For "list_imsi", I think this info we can get it from redis. I added as utility to get which IMSI were found in the log.

So in general, we can use this flow for this cli.

1. Grab one imsi with list_imsi
2. Get the ue_id found from list_ue_id
3. Output with session_trace using imsi + mme_id found in Step 1 and 2.

(We can repeat step 3, with other mme_id to find previous sessions)

Also, we can use this as offline tool :) (without accessing to the AGW). Updated github with some of these comments for clarification.

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
        f = open(self.args.path, "r")
```
```suggestion
        with open(self.args.path, "r") as f:
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
        imsi_pattern = r"\b\d{15}\b"
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
        imsi_list = defaultdict(int)
        lines = self.lines
        for line in lines:
            match = re.findall(imsi_pattern, line)
            if not match:
                continue
            imsi = match[0]
            imsi_list[imsi] += 1
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
        pattern_mme_enb_ue_id = r".*MME_UE_S1AP_ID = (\w+)\b eNB_UE_S1AP_ID = (\w+)\b"
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
            if ue_id and ue_id not in ue_id_list:
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
            if enodeb[1] != "None":
```

**Review by ardzoht — CHANGES_REQUESTED** [review]
left some comments, I suggest running https://docs.magmacore.org/docs/next/lte/dev_unit_testing#format-agw to format the python file

**Comment by wallyrb** [comment]
> left some comments, I suggest running https://docs.magmacore.org/docs/next/lte/dev_unit_testing#format-agw to format the python file

thanks @ardzoht for reviewing. I addressed the comments, run the precommit format script and fix remaining lint errors(with the exception of "WPS306 Found class without a base class: UserTrace" not sure if I can clean that one)

Feel free to add any other suggestions. Thanks!

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
            ue_id_dict= {}
```
for clarity through the code

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]: no need to append `enb_ue_id` here

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]: if we don't append `enb_ue_id` on line 161, we can update the check to:
```suggestion
            mme_ue_id = v.get('mme_id')
            enodeb_ue_id = v.get('enb_ue_id')
            if not enodeb_ue_id:
                print...
            else:
               print...
            
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
            match = re.findall(self.pattern_str, line, re.IGNORECASE)
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
        self.pattern_str = "|".join(pattern_list)
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
    def set_ue_id_pattern(self, args):
```

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]: move this on top of the file and define it as constant
```suggestion
        SESSIOND_ERRORS_LIST = [
```

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (wallyrb) [reply lte/gateway/python/scripts/user_trace_cli.py]: thanks for the suggestion!

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (wallyrb) [reply lte/gateway/python/scripts/user_trace_cli.py]: great suggestion! thank you, I have updated with the comments

**Comment by wallyrb** [comment]
thanks a lot for the suggestions @ardzoht . I have updated the PR

**Review by ardzoht — COMMENTED** [review]
Thanks for the changes Wally, it looks good overall to me, the last part we need to do is to add the script to the `lte/gateway/python/setup.py` under scripts list. Once we have that I will approve the PR

**Comment by wallyrb** [comment]
> Thanks for the changes Wally, it looks good overall to me, the last part we need to do is to add the script to the `lte/gateway/python/setup.py` under scripts list. Once we have that I will approve the PR

thanks @ardzoht will add the script on the list and will add also some documentation for this cli.

**Comment by wallyrb** [comment]
- Added the script under `lte/gateway/python/setup.py`
- Set the mme log default path in the script to /var/log/mme.log
- Add documentation for the script(attached picture in the test plan)
- Some minor changes in description of parameters.

@ardzoht let me know if you have any additional comments. Thanks!

**Review by electronjoe — APPROVED** [review]
`/docs/` CODEOWNER LGTM

**Review by ardzoht — APPROVED** [review]
lgtm

**Review by themarwhal — APPROVED** [review]
This looks great! I'm noticing the GHA jobs are not being triggered on this PR. (Is this the same issue we're seeing with @sebathomas ? cc @quentinDERORY , @electronjoe )

**Comment by sebathomas** [comment]
Github support finally resolved the CI issue over the weekend and explained what happened. Github Actions were disabled entirely for my account because my fork of magma/magma was incorrectly flagged for "mining and reaching tor relay nodes". Unfortunately, they could not explain which part of the workflow was flagged.

If you have the same issue, you should see a red box with an error at the top of the "Actions" page of any repository. The only way to get it fixed is to contact Github support.

**Comment by wallyrb** [comment]
rebasing and fixing markdown lint doc errors that were failing.
Note:  Below file was failing the markdown lint check, fixing to pass this check.
docs/readmes/nms/organizations.md

File: lte/gateway/python/scripts/user_trace_cli.py
Comment (ardzoht) [inline lte/gateway/python/scripts/user_trace_cli.py]:
```suggestion
class UserTrace(object):
```
this will fix remaining pep8 linting error

Comment (wallyrb) [reply lte/gateway/python/scripts/user_trace_cli.py]: Updated, thanks for the suggestion @ardzoht !

**Pull request merged**
